installer-aws: create edge compute pool for AWS Local Zones (#1232)
openshift/enhancements: enhancements/installer/aws-custom-edge-machineset-local-zones.md

change the installer.

The logic of the preferred instance type will be replaced for the `edge` pool when
the user provides the instance type on the install-config, attribute `compute[name=edge].platform.aws.type`.

#### Volume type

The EBS Volume Type `gp3`, used by default on IPI, is not broadly available
across all GA Local Zones locations. [The Local Zones Features page](https://aws.amazon.com/about-aws/global-infrastructure/localzones/features) is the current place to check which location supports `gp3`.

AWS also does not provide an API to query the zone offering for EBS flavors. The EBS
volume type for Local Zones, `edge` compute pool, will be fixed to `gp2`.

The EBS Volume Type can be changed on the `edge` compute pool option
`compute[name=edge].platform.aws.rootVolume.type` on the `install-config.yaml`.

If the user workload requires a Physical Volume, the `storageClass` can be set to `gp2-csi`
on the deployment config.

#### Machine Sets for edge nodes

The installer must generate one MachineSet for each Local Zones location, considering
all the limitations on the Local Zones (instance types, EBS, network, etc).

> The MachineSet for the `edge` pool will also set those custom labels:

Author: > many similar machinesets across many different local zones right?

That's correct, each zone will have one MachineSet.

> But all the taints and configuration for the machinesets will be the same?

Taint yes, will be the same. Labels/configurations, no. The configurations, for example zone-group label, will be different as each Local Zone belongs exclusively to one Zone Group. There will be a case of two zones belonging to the same group (currently the unique example is Los Angeles zones [us-west-2-lax-1{a,b}](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R580-R581)), but the configuration (at least, zoneName) will not be the same either.

https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R493

> The MachineSet for the `edge` pool will also set those custom labels:

Reviewer: Ok, so, might be worth clarifying a bit, if I want to deploy an application, as an end user, to a specific zone at the end, what differentiates my different zones, eg, is there an example, required during scheduling affinity you can show?

> The MachineSet for the `edge` pool will also set those custom labels:

Author: Yes:
- Section [User-workload deployments](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R848-R862)
- Product docs: [Cluster Tasks / Local Zones / Creating user workloads in AWS Local Zones](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/cluster-tasks.html#installation-extend-edge-nodes-aws-local-zones_post-install-cluster-tasks)

> - `node-role.kubernetes.io/edge=''`: used by `taint` and to quickly identify on `oc get nodes`

Reviewer: Is there any implication of it sharing the worker machineconfigpool?

> - `node-role.kubernetes.io/edge=''`: used by `taint` and to quickly identify on `oc get nodes`

Author: I can't see. As a regular worker node, the workers running in the Local Zones should have the same lifecycle as regular workers. As OpenShift and User workloads, it won't have the same lifecycle considering the LZ limitations.
Can you see any restrictions/implications/limitations?

> - `node-role.kubernetes.io/edge=''`: used by `taint` and to quickly identify on `oc get nodes`

Author: One [thread discussed a similar topic](https://github.com/openshift/enhancements/pull/1232#discussion_r980031284)

- https://issues.redhat.com/browse/OSD-14980
- https://issues.redhat.com/browse/OHSS-18215
- https://issues.redhat.com/browse/OCPBUGS-3910
- https://issues.redhat.com/browse/OCPBU-303
- https://issues.redhat.com/browse/SDN-3857

#### Additional IAM permissions

In addition to the resources and permissions required by the regular IPI deployment, the user must have permission to [ModifyAvailabilityZoneGroup](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_ModifyAvailabilityZoneGroup.html).

The permission `ec2:ModifyAvailabilityZoneGroup` must be allowed on the IAM policies attached to the entity (User or IAM Role) used by the agent running the openshift-installer.

#### General considerations

The following items must be satisfied in this phase:

- the installer must allow one item of compute pool type `edge` (`compute[?name=="edge"]`)

- the installer must extract Local Zone subnets from the subnet IDs
  on `platform.aws.subnets`, creating one MachineSet manifest for each zone.

- the `compute[?name=="edge"]` must create MachineSets manifests on Local Zone
  subnets with:
  - labels: `node-role.kubernetes.io/edge=''`, `machine.openshift.io/zone-type=local-zone`, and

>     `machine.openshift.io/zone-group=zone_group_name`;

Author: Those node `labels` can be used by users to create their workloads.

https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R848-R862

https://docs.openshift.com/container-platform/4.13/post_installation_configuration/cluster-tasks.html#installation-extend-edge-nodes-aws-local-zones_post-install-cluster-tasks00

>     `machine.openshift.io/zone-group=zone_group_name`;

Reviewer 2: Machine labels automatically get copied as node labels?

>     `machine.openshift.io/zone-group=zone_group_name`;

Author: Yes: https://github.com/openshift/machine-api-operator/blob/master/FAQ.md#adding-annotations-and-labels-to-nodes-via-machines

  - taint `NoSchedule` matching the label `node-role.kubernetes.io/edge`

- the MachineSet should be created when any Local Zones subnet is added on

>   `platform.aws.subnets`, even if the compute pool was not defined on `compute[?name=="edge"]`

Author: We chatted about that topic [here](https://github.com/openshift/enhancements/pull/1232#discussion_r979985450).

To install a cluster in an existing VPC you must provide the subnets. The installer will assume this path when `platform.aws.subnets` is not empty [default installer behavior].

>   `platform.aws.subnets`, even if the compute pool was not defined on `compute[?name=="edge"]`

Reviewer: I clearly don't remember things I've said before, there's a link at the end that shows where you called this out but it's broken now, do you have a pointer to where you called this out?

>   `platform.aws.subnets`, even if the compute pool was not defined on `compute[?name=="edge"]`

Author: > there's a link at the end that shows where you called this out but it's broken now

Oh no! thanks for getting this. I will fix it now. The direct link of the discussion is: 

https://github.com/openshift/enhancements/pull/1232#discussion_r1045918478

>   `platform.aws.subnets`, even if the compute pool was not defined on `compute[?name=="edge"]`

Author: I will append a description of the existing flow when installing in exist VPC, so we can make sure that information/flow is not new or something introduced in this EP:
```
- the MachineSet should be created when any Local Zones subnet is added on
  `platform.aws.subnets`, even if the compute pool was not defined on `compute[?name=="edge"]`.
  That behavior is aligned with the same installer flow of `worker` compute pool when
  installing a cluster in existing network/VPC: subnet(s) created in the zone type
  `availability-zone` is provided.
```

---

> title: aws-custom-edge-machineset-local-zones

Author: @patrickdillon @JoelSpeed @makentenza @rvanderp3 could you please take a look at this EP?

Author: Also @dhellmann, Thanks!

Author: @elmiko would be happy with your thoughts here too. Thanks! =]

Reviewer 3: @mtulio sorry, just saw this ping. will take a read

Phase 2:

1. The Local Zones currently have limitations of Nat Gateways, which keep the gateway
  strategy more complex to provide egress traffic for compute nodes deployed on
  that zone. Can we keep the edge nodes in the public subnet? Do we need to
  create a dedicated security group instead of using/sharing with the default worker
  machine pool?

1. How we can keep extending the Zones for eventually new locations? AWS has a fixed
  number of Local Zones available, when we create a cluster we will try to calculate
  the subnets IPs to cover all IP ranges allowed by VPC (/16) depending on the number
  of subnets available. I also wonder if we can consider a smaller range of IPs for
  subnets operating in Local Zones, allowing the customer to grow it later. Can we
  consider using a lower CIDR for subnets hosted on Local Zone locations?

1. How we can advise customers to grow locations Day-2? Can we keep a product
  documentation page with that instructions? A sort of instructions on how to create
  CloudFormation Stack to create a new zone for the new location that is available, or
  the customer wants to support in the future.

### Test Plan

> TBD

Author: Check:
https://github.com/openshift/release/blob/master/ci-operator/step-registry/upi/install/aws/cluster/upi-install-aws-cluster-commands.sh

also have limited and more expensive cloud services, cluster operators should be
able to create worker nodes on the edge networks on day zero operations.

This enhancement proposes a new compute pool named `edge`, which will create
Machine Sets for each availability zone on the AWS Local Zones to provision compute
resources with `NoSchedule` taint.

The EP is divided into three phases:

- Phase 0: Research. Understand the Local Zones using the OKD/OCP, create use case, and
  provides the documentation of the installation of a cluster in an existing VPC.
- Phase 1: Install a cluster in user provided network introducing to the installer a new
  compute pool named `edge`. The MachineSet manifests will be created for each subnet
  provided on `platform.aws.subnets` for Local Zones. The Machine will be protected with
  taints to avoid running the normal platform workloads.
- Phase 2: The installer should automate the network resource creation when the user
  provides zone names with the type `local-zone` on the `platform.aws.zones` or
  `compute[?name=="edge"].platform.aws.zones`.

Phases 1 and 2 are covered by this enhancement.

## Motivation

> Presently, the installer does not support Local Zone subnets on day zero operations

Author: > I'm interested to know how hard it is to set up these edge nodes as a day 2 operation, what's involved?

The steps are like this:
- check if the VPC has enough CIDR blocks to add new subnets
- opt-in zone group
- create subnets on the Local Zone, setting the Tags to avoid the ingress controller discovery of this subnet
- associate the subnet with a public route table
- check the instance type supported on the AZ
- create the MachineSet for each Local Zone AZ, setting the taints for that MachineSet (for reasons presented on this EP on why not sharing the nodes on the edge with normal workloads)
- scale the machines

If the customer would like to deploy in the private subnet, it could become more complex as it needs to use the Nat GW from the parent region.

If the VPC haven't enough blocks, the customer will not be able to use Local Zones.

> Where has the need for day 0 come from? Could it be/would it be simpler to have day 2 set up only?

[RFE-2782](https://issues.redhat.com/browse/RFE-2782). Mainly telecom customers may be able to fully deploy workloads on the edge without extra effort or hacking on Day-2, or Day-0 adding extra manifests or installing in existing VPC. The second phase will implement support to installer create all required resources.
cc @makentenza could you please add your points of customer requesting that feature?

> Presently, the installer does not support Local Zone subnets on day zero operations

Reviewer: Both the "what does this look like today" flow and the motivation that we have customers looking for 1 touch installations across many edges are useful contexts that would be good to capture in the enhancement IMO

> Presently, the installer does not support Local Zone subnets on day zero operations

Author: @JoelSpeed thanks for raising it. The ["Motivation"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R60-R97) has been updated with your feedback on this thread.

for those reasons:
- the Local Zone does not support network load balancers (NLB or CLB), used by
  ingress deployed on worker pool.
- the Local Zone does not support Nat Gateway, used as a gateway on the default IPI
  deployment for each subnet and availability zone.
- the default instance type for compute nodes used on IPI, `m6i.xlarge`, is not
  available on Local Zones
- the default EBS type for IPI, `gp3`, is not available in the most Local Zone locations
- segment the `worker` compute node pool is not allowed

The resources on Local Zones were designed for applications delivering low latency to
the end-users, in general, hosted in the data plane nodes. Furthermore, the price of
cloud services is more expensive, and the services are limited. Finally, mixing the
ingress router backends from different locations in the Local Zones can impact
drastically, and directly, the end users' experience (ex: routing requests from users
from NYC to node hosted in Local Zones in Miami, instead of New Year node).

> As the applications using edge locations are very specific workloads, cluster admins

Reviewer 4: Maybe note this is reason for `NoSchedule`

Phase 1:

- As a cluster admin, I want the ability to specify a set of subnets on the AWS
  Local Zone locations to deploy worker nodes, so I can create custom applications
  to deliver low latency to my end users.

- As a cluster admin, I would like to create a cluster extending worker nodes to the
  edge of the AWS cloud provider with Local Zones, so I can create custom applications
  to deliver low latency to my end users.

- As a cluster admin, I would like to control the workloads running on the edge
  compute pool, and keep it dedicated for applications which are latency sensitive,
  so I can have a cost-effective infrastructure, avoiding mixing workloads in different locations.

- As a cluster admin, I would like to select existing subnets from the local and
  the parent region zones, to install a cluster, so I can manage my custom networks with
  my automation.

Phase 2:

- As a cluster admin, I would like to install OpenShift clusters, extending the
  compute nodes to the Local Zones in my day zero operations without needing to
  set up the network and compute dependencies, so I can speed up the edge adoption
  in my organization using OKD/OCP.

### Goals

Phase 1:

> - Admins can create compute pool named `edge` on the AWS platform to setting up Local Zone MachineSets.

Reviewer 5: This wording is a bit confusing to me. Are you talking about specifying a machine pool in the install config or having users pre-provision infrastructure? Not sure what exactly is meant by "compute pool."

Author: Thanks for your feedback, @patrickdillon . 
> specifying a machine pool in the install config

This. The goal is to allow to installer allow to create an extra pool on install config named `edge` when installing in existing VPC with Local Zone subnets. The installer will automate the creation of MachineSets for each subnet's zone located on Local Zones.

I would re-phrase it to something like this:
- Admins can create machine pools named `edge` when installing a cluster in existing VPC with subnets in AWS Local Zones.

what do you think?

Author: Item [reviewed](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R131).

  connectivity between the compute nodes and the carrier network.

Phase 1:

- Create the network resources on the edge locations. That implementation will be
  done in the second stage, when the installer will fully automate the creation
  of resources on the availability zones specified on the `compute[.name=='edge'].platform.aws.zones`.

## Proposal

The phases proposed in this enhancement are:

- Phase 1: Introduce the `edge` compute pool restricted to AWS Local Zones. Allow
  users to install a cluster in existing networks with Local Zone subnets.
- Phase 2: Create the network resources for AWS Local Zone locations.

### Phase 1: Introduce the "edge" compute pool

The first phase will introduce the new group of compute pool designed to
be used exclusively for the user workloads on the zones not hosted in the
default availability zones within the region, where the cluster is used to
be deployed.

Due to the limitation of cloud services in the edge locations, the dedicated
pool also has limited compute and EBS type when compared to the default

> IPI (i.e. on AWS: m6i.xlarge).

Author: > Do these edge locations ever get newer instance types?

Not newest, most of them do not have the 6th-gen instances, currently the latest generation used by the installer.

> wondering if we will have similar issues with these edge zones where we will have to keep them up to date.

It will follow the same approach as nowadays. See my comment with regards to how the installer chooses the type based on it's [zone availability](https://github.com/openshift/enhancements/pull/1232/files#r979172587). The `edge` pool will use the same engine, but will not impact the decision of the current worker or master pool instance types.

> Is there a consistent set of instance types across all the edge zones that we can use as a default choice?

We can query the [instance offerings](https://aws.amazon.com/ec2/instance-types/) to check it. I exported it across locations between us-east-1 and us-west-2, take a look on the list:
https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42

As described on this EP, it seems we could use the existing installer preferred flow appending the `c5d.2xlarge` to cover all Local Zone locations, then the evaluation order will be: m6i.xlarge, m5.xlarge, and c5d.2xlarge.

> IPI (i.e. on AWS: m6i.xlarge).

Author: @JoelSpeed Thanks for your feedback. Revisiting this topic I can see one situation that could be one issue when defining a unique instance type for the entire pool (that's the only exception I can see):
- When deploying a cluster in `us-west-2` supporting `edge` pools on subnets `us-west-2-lax-1{a,b}` [Los Angeles] AND any other Local Zone location within `us-west-2` (let's use `us-west-2-las-1a` [Las Vegas])

With that scenario, the installer will lookup[1] for an intersection of instance types from the default Instances over **all AZs for the pool**. It will not find any, as `c5d.2xlarge` is not offered in any zone of Los Angeles, AND `m5.xlarge` is not offered in any zone out of Los Angeles. Finally, the default preferred type `m6.xlarge` will be choosed. It will result in the Failed machine when the cluster will be created. The Los Angeles location is the unique exception that does not offer `c5d.2xlarge` along all zones previously queried[2].

[1] https://github.com/openshift/installer/blob/master/pkg/asset/machines/aws/instance_types.go#L17-L39
[2] https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42

Said that, and considering that those pools will not run the same workload due to the taints, I propose to use the same preferred engine to choose the instance type to check the preferred order **by AZ**, as we are always creating MachineSet by AZ.

So in this case, given the new preferred list for those zones, the installer will generate those MachineSets:

```
>> 99_openshift-cluster-api_worker-machineset-0.yaml
replicas	: 1
AZ		: us-west-2a
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-1.yaml
replicas	: 1
AZ		: us-west-2b
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-2.yaml
replicas	: 1
AZ		: us-west-2c
instanceID	: m6i.xlarge
EBS		: gp3
zone_type	 null
meta		: {}

>> 99_openshift-cluster-api_worker-machineset-3.yaml
replicas	: 1
AZ		: us-west-2-lax-1a
instanceID	: m5.xlarge
EBS		: gp2
zone_type	 local-zone
meta		: {
  "labels": {
    "node-role.kubernetes.io/edge": "",
    "zone_group": "us-west-2-lax-1",
    "zone_type": "local-zone"
  }
}

>> 99_openshift-cluster-api_worker-machineset-4.yaml
replicas	: 1
AZ		: us-west-2-las-1a
instanceID	: c5d.2xlarge
EBS		: gp2
zone_type	 local-zone
meta		: {
  "labels": {
    "node-role.kubernetes.io/edge": "",
    "zone_group": "us-west-2-las-1",
    "zone_type": "local-zone"
  }
}

```

The "best/preferred" instances will be choosed chosen for each AZ: `m5.xlarge` for `us-west-2-lax-1a`, and `c5d.2xlarge` for `us-west-2-las-1a`

Thoughts?

> IPI (i.e. on AWS: m6i.xlarge).

Author: @JoelSpeed  could you please elaborate "we want to be more explicit about it"?

The logic proposed is to select automatically when the `compute[name==edge].platform.aws.type` will not be set, otherwise it will have more precedence. That strategy follows the current flow for `worker` machine pool. Is it make sense to you?

> IPI (i.e. on AWS: m6i.xlarge).

Reviewer: Well I'm wondering if using this logic, there's a possibility that different worker zones might get different instance types and different edge zones may get different types from one another when they don't share.

I would expect the logic to go through each worker zone and check the available types and choose an instance type that is available across them all, and then repeat that process with the edge zones

Unless we are happy with edge foo being c5d.2xlarge and edge bar being m5.xlarge, would be odd for the same worker types to default to different instance types right?

> IPI (i.e. on AWS: m6i.xlarge).

Author: > (1) Well I'm wondering if using this logic, there's a possibility that different worker zones might get different instance types
> (2) and different edge zones may get different types from one another when they don't share.

(1) it will not happen. The current implementation will not be changed, preserving the current flow: the `worker` gets the same instance type across all regular zones.
(2) it will happen in this proposal. The proposal is to use the existing flow, get the preferred instance, looking to the scope of zone, instead of all zones, when choosing the `edge` instances. Example output on us-west-2 cluster deployment is available [here](https://github.com/openshift/installer/pull/6371#issuecomment-1260409687).

> I would expect the logic to go through each worker zone and check the available types and choose an instance type that is available across them all, and then repeat that process with the edge zones

Is it the idea with this logic of lookup of type offerings between all zones (regular and local) set the same instance between `worker` and `edge`?

If yes, I wonder if one of those problems will drive the solution more complex:
- the cluster costs will be increase (2x) as we will choose a common instance to use across all AZs: Currently, the intersection type between AZs is `c5d.2xlarge` (according to the [list exported above](https://gist.github.com/mtulio/c98aa15128a7becb06a372f00d824c42)), which costs 2x the `m6i.xlarge`
- the type `c5d.2xlarge` provides an ephemeral disk, which is good when there's a process to mount and use it by default. It's not the case with OCP, so we will not use this resource, which makes that instance type ~12% more expensive when compared with the same family without that storage (`c5.2xlarge`).
- we may need to mark some Local Zones as unused: `us-west-2-lax-1` zone group does not have "intersection" instance types between all zones. This zone was the first available on Local Zones, currently the unique zone group with two AZs in LZ, and the most modern one providing cheaper instances (`m5.xlarge`), and also `gp3` storage.
- we may change the entire cluster types when using a new feature (Local Zones/edge pool), which may increase the complexity in Support of changing the default instance types for worker and control plane nodes, instead of the expected `m6i.xlarge`.

> Unless we are happy with edge foo being c5d.2xlarge and edge bar being m5.xlarge, would be odd for the same worker types to default to different instance types right?

I am not sure if we would need to be concerned about different instance types between `worker` and `edge`, and within `edge` zone groups for those reasons:
- `worker` should not, initially, share the same workloads as `edge`
- nodes between different `edge` groups may not share the same workload. For that reason, we are setting taints and custom labels by zoneGroup to give control to admins/developers to choose what workloads to run for each edge location. For example, if the administrator creates a new ingress (ALB) sharing nodes from different Local Zones zone groups (i.e.: LAS and LAX), their end-users will have different experiences (sometimes worst) in terms of network latency.
- we are using a "best instance" strategy for the `edge` pools, decreasing the necessity of maintenance on the code
- we are still using the standard types from the default IPI (without edge/Local Zones support)

@JoelSpeed Thoughts? What do you think about those points?

> IPI (i.e. on AWS: m6i.xlarge).

Author: I also would love to hear the ideas of @patrickdillon @makentenza  @dhellmann and @sdodson

> IPI (i.e. on AWS: m6i.xlarge).

Reviewer: > Is it the idea with this logic of lookup of type offerings between all zones (regular and local) set the same instance between worker and edge?

No. I would expect all workers to be the same, and all edge to be the same, if edge and worker differ I think that's ok

> nodes between different edge groups may not share the same workload. For that reason, we are setting taints and custom labels by zoneGroup to give control to admins/developers to choose what workloads to run for each edge location. For example, if the administrator creates a new ingress (ALB) sharing nodes from different Local Zones zone groups (i.e.: LAS and LAX), their end-users will have different experiences (sometimes worst) in terms of network latency.

This is a fair point, I don't particularly have a strong opinion on this

> IPI (i.e. on AWS: m6i.xlarge).

Reviewer: Yeah that's really interesting and helps argue the case for edge groups being completely independent of one another, if you can include a bit about this that would be cool, thank you!

> IPI (i.e. on AWS: m6i.xlarge).

Author: @JoelSpeed section added with a summary of the benchmark mentioned above - which led to the proposal of allowing different types on the edge pools from different zone groups:
https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R698-R769

Thanks!

The edge machine pool will be available only when deploying on the AWS platform
with Local Zones. In the future, it can be extended to the AWS Wavelength,
and other cloud providers if needed.

The following items must be satisfied in this phase:

- the installer must allow one item of machine pool type `edge` (`compute[?name=="edge"]`)
- the installer must extract Local Zone subnets from the subnet IDs
  on `platform.aws.subnets`, creating one MachineSet manifest for each zone.
- the `compute[?name=="edge"]` must create MachineSets manifests on Local Zone
  subnets with:
  - labels: `node-role.kubernetes.io/edge=''`, `location=local-zone`, and
    `zone_group=zone_group_name`
  - and taint `NoSchedule` matching the label `node-role.kubernetes.io/edge`
- the MachineSet should be created when any Local Zone subnet is added on

>   `platform.aws.subnets`, even if compute pool was not defined on `compute[?name=="edge"]`

Author: > Why "even if"?

To keep compatibility with the existing flow when installing in existing VPC. The installer will look at each ZoneType of subnetIDs provided on `platform.aws.subnets`.

> Surely we are asking the user to specify the definition for the edge in compute[edge]?

Phase 1 covers only installation in existing VPC, thus the unique requirement is to provide the subnetID located in Local Zones. The user can provide the edge pool definition if they want to customize the pool, but may not be required. Thoughts?

>  How do we work out if platform.aws.subnets contains a local zone subnet?

The ZoneType attribute for each subnetID. [1]

> Is that something that is easy to look up?

Yes, currently there's a classification of ["public" or "private"](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/subnet.go#L113-L121) subnets, the edge lookup will be added in the same flow.

> Feels like we are creating an implicit behaviour rather than being explicit about it

IIUC, the 'implicit behavior only' may not keep the compatibility with existing installer flow when installing in existing VPC, where it expects to create MachineSets on all AZs discovered by subnetID.

[1] https://github.com/openshift/enhancements/pull/1232/files#r979171225

>   `platform.aws.subnets`, even if compute pool was not defined on `compute[?name=="edge"]`

Reviewer: > To keep compatibility with the existing flow when installing in existing VPC. The installer will look at each ZoneType of subnetIDs provided on platform.aws.subnets.

That should be called out, it wasn't obvious to me that this provides backwards compatibility :)

>   `platform.aws.subnets`, even if compute pool was not defined on `compute[?name=="edge"]`

Author: I created new sections under the [Proposal of Phase-1](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R178) to cover your items here. Thanks for sharing it. :)

> Note: Local Zones subnet scope. See section "Public subnet and outbound traffic"
- only `public` subnets are allowed for Local Zone subnets ID added on `platform.aws.subnets`
- the Local Zone type should not be allowed on the `worker` compute pool
  `compute[?name=="worker"].platform.aws.zones`.

- the default or user-provided (`compute[?name=="edge"].platform.aws.type` or
  `platform.aws.defaultMachinePlatform.type`) instance type for the machine
  pool `compute[?name=="edge"]` should be validated on the EC2 offering for the
  target availability zone

> > Note: Currently the resource types are limited. The intersection between zones is c5d.2xlarge and gp2

Author: The instance types can be queried by the API [DescribeInstanceTypeOfferings](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTypeOfferings.html). The EBS type can is available on [Local Zone Features](https://aws.amazon.com/about-aws/global-infrastructure/localzones/features) page.

The installer will query it when creating the manifests. If the user would like to provide it, he needs to lookup up the AWS documentation or queries the API, I think.

I will update that note with those references.

> > Note: Currently the resource types are limited. The intersection between zones is c5d.2xlarge and gp2

Author: Links and references [have been added](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R196-R225) to the sections (instance and volume) of the Proposal. Thanks!

- the default `edge` machine pool instance type must be common in the current
  availability zone offerings. The default for the `edge` pool will be `c5d.2xlarge`.
- the default `edge` machine pool EBS type must be common in the current
  availability zone offerings. The default for the `edge` pool will be `gp2`.

#### Examples configuration for user-provided network

Those are two examples of a user-provided network when installing a cluster
in existing VPC:

> 1) `install-config.yaml` with Local Zone subnets, and empty `edge` machine pool

Author: The explanation described here was moved to a section ["subnet selection for the edge machine pool"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R178-R194). Thanks for raising it.

```yaml
apiVersion: v1
baseDomain: example.com
metadata:
  name: edge-cluster
platform:
  aws:
    region: us-east-1
    subnets:
    - subnetID-us-east-1a
    - subnetID-us-east-1b
    - subnetID-us-east-1c
    - subnetID-us-east-1-nyc-1a
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

2) `install-config.yaml` with Local Zone subnets, and custom instance type

> Note: the instance type `c5d.2xlarge` common between locations

```yaml
apiVersion: v1
baseDomain: example.com
metadata:
  name: edge-cluster
compute:
- name: edge
  platform:
    aws:

>       type: c5d.2xlarge

Author: ```
$ oc get machines -n openshift-machine-api 
NAME                                               PHASE     TYPE         REGION      ZONE               AGE
ipi-local-zones-tzzwt-edge-us-west-2-lax-1a-7w8d8  Running   m5.xlarge    us-west-2   us-west-2-lax-1a   36m
ipi-local-zones-tzzwt-edge-us-west-2-lax-1b-26tpq  Running   m5.xlarge    us-west-2   us-west-2-lax-1b   36m
ipi-local-zones-tzzwt-master-0                     Running   m6i.xlarge   us-west-2   us-west-2b         40m
ipi-local-zones-tzzwt-worker-us-west-2a-wgklg      Running   m6i.xlarge   us-west-2   us-west-2a         36m
```

platform:
  aws:
    region: us-east-1
    subnets:
    - subnetID-us-east-1a
    - subnetID-us-east-1b
    - subnetID-us-east-1c
    - subnetID-us-east-1-nyc-1a
    - subnetID-us-east-1-mia-1a
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

3) `install-config.yaml` with Local Zone subnets, and custom volume type

> > Note: currently only `us-west-2` supports `gp3`

Reviewer 6: Is this exposed through an API so we can use it to validate this kind of restriction beforehand?

Author: This topic was moved to a section ["Volume type"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R216). Thanks!

```yaml
apiVersion: v1
baseDomain: example.com
metadata:
  name: edge-cluster
compute:
- name: edge
  platform:
    aws:
      rootVolume:
        type: gp3
platform:
  aws:
    region: us-west-2
    subnets:
    - subnetID-us-west-2a
    - subnetID-us-west-2b
    - subnetID-us-west-2c
    - subnetID-us-west-2-lax-1a
    - subnetID-us-west-2-lax-1b
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

> The `install-config.yaml` exemplified above will generate oen MachineSet

Reviewer 4: ```suggestion
The `install-config.yaml` exemplified above will generate one MachineSet
```

by subnet ID on install directory `<install-dir>/openshift/`:

> Example  `openshift/99_openshift-cluster-api_worker-machineset-3.yaml`:

Reviewer 6: Would it make sense to have the compute pool name "edge" in the file name here? E.g.

```
openshift/99_openshift-cluster-api_edge-machineset-3.yaml`
```

Author: We could, but it's only for the installer manifests. The CR will have the `edge` on the name instead of `worker`:
```
$ oc get machinesets -n openshift-machine-api 
NAME                                         DESIRED   CURRENT   READY   AVAILABLE   AGE
ipi-local-zones-tzzwt-edge-us-west-2-lax-1a  3         3         3       3           34m
ipi-local-zones-tzzwt-edge-us-west-2-lax-1b  2         2         2       2           35m
ipi-local-zones-tzzwt-worker-us-west-2a      1         1         1       1           35m
ipi-local-zones-tzzwt-worker-us-west-2b      1         1         1       1           35m
ipi-local-zones-tzzwt-worker-us-west-2c      1         1         1       1           35m
```

```yaml
apiVersion: machine.openshift.io/v1beta1
kind: MachineSet
metadata:
  labels:
    machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
  name: ${CLUSTER_ID}-edge-us-east-1-nyc-1a  (1)
  namespace: openshift-machine-api
spec:
  replicas: 1
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
      machine.openshift.io/cluster-api-machineset: ${CLUSTER_ID}-edge-us-east-1-nyc-1a (1)
  template:
    metadata:
      labels:
        machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
        machine.openshift.io/cluster-api-machine-role: edge  (2)

>         machine.openshift.io/cluster-api-machine-type: edge  (2)

Reviewer: I wonder how many places have semantics about valid values for these labels 🤔 Currently it is Infra, Worker, Master and that should be it?

Reviewer: Cool that makes sense, might still be nice to have the edge role on the machines though, we do the same for infra so I don't see why we shouldn't be consistent with that here

Reviewer: My comment was more along the lines of, we should really try this out to see if anything starts complaining

Author: @JoelSpeed [this is the Installer PR](https://github.com/openshift/installer/pull/6371#issuecomment-1260409687), the machines were created with the edge role though  - your suggestions above.
(also I appreciate if you would like to take a look/review/feedback there =) )

Reviewer: Looks like it should work with the edge role

          instanceType: c5d.2xlarge
          kind: AWSMachineProviderConfig
          placement:
            availabilityZone: us-east-1-nyc-1a (7)
            region: us-east-1
          securityGroups:
          - filters:
            - name: tag:Name
              values:
              - ${CLUSTER_ID}-worker-sg
          subnet:
            id: subnetID-us-east-1-nyc-1a (8)
          publicIp: true  (9)
          tags:
          - name: kubernetes.io/cluster/${CLUSTER_ID}
            value: owned
          userDataSecret:
            name: worker-user-data
```

- 1) The `worker` part of the MachineSet name will be replaced with `edge`.
- 2) MachineSet labels for the `edge` compute nodes.
- 3) The label `zone_type` is the type of the availability zones. The [API attribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html) is `ZoneType`.
- 4) The label `zone_group` is the group to which the availability zone belongs. The [API attribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html) attribute is `NetworkBorderGroup`.
- 5) The `NoSchedule` taint will be added for each node with the

>   tag `node-role.kubernetes.io/edge=''`.

Reviewer 6: Is this already recognized by Kube?

Author: Yes, it is. the `edge` node role will coexist with the current one `node-role.kubernetes.io/worker=''`. The idea to use that tag and label is to have a good visualization when describing the nodes:

```
 oc get nodes
NAME                                         STATUS   ROLES                  AGE   VERSION
ip-10-0-12-34.us-west-2.compute.internal     Ready    control-plane,master   32m   v1.24.0+8c7c967
ip-10-0-13-194.us-west-2.compute.internal    Ready    worker                 24m   v1.24.0+8c7c967
ip-10-0-17-18.us-west-2.compute.internal     Ready    control-plane,master   32m   v1.24.0+8c7c967
ip-10-0-17-88.us-west-2.compute.internal     Ready    worker                 14m   v1.24.0+8c7c967
ip-10-0-192-236.us-west-2.compute.internal   Ready    edge,worker            22m   v1.24.0+8c7c967
ip-10-0-194-18.us-west-2.compute.internal    Ready    edge,worker            23m   v1.24.0+8c7c967
(...)
```

Reviewer: I wonder if MCO supports booting directly into custom machine config pools yet, if they do, you wouldn't have to have the worker role on these which would make things cleaner

Author: I am not sure, but I think if we would like to use custom MCP we may need to customize/duplicate MachineConfigs which will generate more manifests to manage and more changes on the installer, and I am not sure if could impact the lifecycle. I thought to use that one as it seems the [same approach used](https://docs.openshift.com/container-platform/4.11/machine_management/creating-infrastructure-machinesets.html#assigning-machineset-resources-to-infra-nodes) for `infra` nodes. We can use another label than `node-role.kubernetes.io/edge`, but it may be more difficult for the cluster operator/admins to identify `edge` instances on the `oc get nodes`.

Reviewer: I think the approach here is reasonable so I don't see a need to change it. In terms of duplication, I believe you can extend machine configs so i don't think it would be too bad, maybe 2 or 3 additional files. But you're right, we already have precedence for this pattern with Infra so lets keep it the way it is right now

Author: Thanks for the feedback from you all. I created [one new section](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R228-R243) detailing what we discussed here to keep things more clear.

Reviewer 5: > Is this already recognized by Kube?

> Yes, it is. the edge node role will coexist with the current one node-role.kubernetes.io/worker=''

This is interesting, are there any more details to elaborate on here? That is, is the node-role.kubernetes.io/edge='' specifically recognized by k8s and is there reference to this?

Author: > > Is this already recognized by Kube?
> 
> > Yes, it is. the edge node role will coexist with the current one node-role.kubernetes.io/worker=''
> 
> This is interesting, are there any more details to elaborate on here? That is, is the node-role.kubernetes.io/edge='' specifically recognized by k8s and is there reference to this?

I answered `recognized` as `allowed to set that label`.

According to the official documentation, only `control-plane` is a [well-known label documented on the k8s](https://kubernetes.io/docs/reference/labels-annotations-taints/). (@JoelSpeed feel free to correct me on that)

Why I proposed `node-role.kubernetes.io/edge` **to coexist** `node-role.kubernetes.io/worker`? For UX. Users can easily identify which nodes are `edge` or regular in `oc get nodes`. We could set the taints in any of the other labels [introduced in this EP](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R241-R243), example `zone_group` without `node-role/edge`. As mentioned in my comment above, it is also a well-known approach used by `infra` "pool" in OCP, so I thought why not!

Why I didn't propose only `node-role.kubernetes.io/edge`? It will become the solution more complex, setting extra MachineConfigs, setting up kubelet, etc (I didn't map how many changes to do when choosing this path). But I thought as over scoped.

baseDomain: example.com
metadata:
  name: edge-cluster
compute:
- name: edge
  platform:
    aws:
      zones:
      - us-east-1-nyc-1a
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

1) example customizing the instance and EBS type

```yaml
apiVersion: v1
baseDomain: example.com
metadata:
  name: edge-cluster
compute:
- name: edge
  platform:
    aws:
      type: m5.xlarge
      rootVolume:
        type: gp3
      zones:
      - us-west-2-lax-1a
      - us-west-2-lax-1b
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

The following MachineSet will be created for the example 1:

```yaml
apiVersion: machine.openshift.io/v1beta1
kind: MachineSet
metadata:
  labels:
    machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
  name: ${CLUSTER_ID}-edge-us-east-1-nyc-1a
  namespace: openshift-machine-api
spec:
  replicas: 1
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
      machine.openshift.io/cluster-api-machineset: ${CLUSTER_ID}-edge-us-east-1-nyc-1a
  template:
    metadata:
      labels:
        machine.openshift.io/cluster-api-cluster: ${CLUSTER_ID}
        machine.openshift.io/cluster-api-machine-role: edge

>         machine.openshift.io/cluster-api-machine-type: edge

Author: ignore it according Joel's feedback.

- 1 compute edge node on the public subnet in Local Zone

### Workflow Description

TBD

#### Variation [optional]

<!-- If the cluster creator uses a standing desk, in step 1 above they can
stand instead of sitting down.  

See
https://github.com/openshift/enhancements/blob/master/enhancements/workload-partitioning/management-workload-partitioning.md#high-level-end-to-end-workflow
and https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/automated-workflow-for-agent-based-installer.md for more detailed examples. -->

TBD

### API Extensions

Not applied. No new APIs will be added.

### Implementation Details/Notes/Constraints [optional]

#### Public subnet and outbound traffic

It is a best practice to place the worker nodes in private subnets, although the
AWS provides limited services in the Local Zones requiring a more complex setup to do so.

Those are some options to create subnets on Local Zones:

1) Create a subnet and associate it to a route table with the internet gateway as default.
2) Create a private subnet, associating it to a route table from the zone in the
  parent region. The cost will be twice as the traffic should cross the zone to

>   access the internet through a Nat Gateway, which also has fees to be applied.

Reviewer: Have we measured the cost of adding an edge worker? Kubelet to Kube API traffic is pretty intense so I wonder what that's going to cost per node?

Author: The internal (within the VPC: kubelet<->NLBs) data transfer costs will be the same of the other (normal workers) nodes[1][2]. The compute resources in Local Zones is more expensive, for example `m5.xlarge` hosted in the parent zones of `us-west-2` will cost $0.192/hr, and in the LAX (Los Angeles) Local Zone is $0.23/hr.

The scenario described in this specific line is when following the architecture of using private subnets in Local Zone re-using Nat Gateway from the parent region, as NAT GW is not supported on the Local Zones locations.

In this case, the following fees will be applied when the node needs access to the internet:
- Fee for data transferred "in" to and "out" from NAT GW located in another AZ ($0.01/GB for each direction)
- NAT GW fee per GB processed $0.045

There is another cost of $0.045/hr for NAT GW allocation. But this is not exclusive to a node in LZ, as we can assume that the Nat GW is already created to be used in a valid AZ.

When the node is in the public subnet (described in this EP), the fees applied will be the same to access the internet.

[1] [EC2 Data transfer fees](https://aws.amazon.com/ec2/pricing/on-demand/)
[2] [EC2 Local Zones Pricing](https://aws.amazon.com/about-aws/global-infrastructure/localzones/pricing/):
> Data Transfer charges in AWS Local Zones is the same as in the Availability Zones today. Learn more about data transfer prices “in” and “out” of Amazon EC2 here.

Author: > So is this to say a local zone instance doesn't actually cost significantly more (aside from the instance type charges) than a regular az instance?

Yes, will not cost significantly more. The EC2 costs are fixed by location. In the case mentioned above, it was 17% more expensive.

> The data transfer costs aren't actually different?

No. The reference[2] from my last comment can confirm it. It was copied from the AWS Documentation page.

Reviewer: Cool, this is useful context, can we make sure this is noted somewhere

Author: I created a [new section "Infrastructure costs"](https://github.com/openshift/enhancements/pull/1232/commits/fdbdba4c38f3b17c5159b4f5206a6a3671bc74ce#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R620-R645) explaining what we discussed here. Thanks for the feedback.

Group must not expose any ports to the internet. This must be the default behavior,
and the security group for the `worker` pool will be used on the `edge` pool.

> TODO: check if the current deployment is not exposing any port to the `0.0.0.0/0` on `worker` SG

#### User-workload deployments

Application developers must set the tolerations for the application running for each location.

The pod spec example below describes how to deploy one application on the node on the zone group `ZONE_GROUP`:

```yaml
      nodeSelector:
        zone_group: ZONE_GROUP
      tolerations:
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
```

<!--
> TODO create one example of deploying the application on the same ZONE_GROUP, spreading it into different zones.

> Currently only Los Angeles (LAX) location has more than one zone[1];

The example below shows how to deploy one application with two replicas distributed
between two Local Zones[1] in the same zone group[2].

[1] Local Zones available in Los Angeles `us-west-2-lax-1a` and `us-west-2-lax-1b`
[2] Local Zone group of Los Angeles `us-west-2-lax-1` -->

> #### User-workload ingress traffic

Reviewer 6: This section might have to be in the docs. ;)

Author: [x-ref this thread](https://github.com/openshift/installer/pull/6371/files#diff-affa4217d0878d1b05bbd3a78a4e9dbbb537dfeb5fb45674aba3646655201fe0R406-R420). Thanks! =)

To expose the applications to the internet on AWS Local Zones, application developers
must expose their applications using AWS Application Load Balancers (ALB). The
[ALB Operator](https://docs.openshift.com/container-platform/4.11/networking/aws_load_balancer_operator/install-aws-load-balancer-operator.html) is available through OLM on 4.11+.

To explore the best of deploying applications on the edge locations, at least one new
ALB `Ingress` must be provisioned by location to expose the services deployed on the
zones.

If the cluster-admin decides to share the ALB `Ingress` subnets between different locations,
it will impact drastically the latency for the end-users when the traffic is routed to
backends (compute nodes) placed in different zones that the traffic entered by Ingress.

The ALB deployment is not covered by this enhancement.

### Risks and Mitigations

TBD

### Drawbacks

TBD

## Design Details

### Open Questions [optional]

> 1. Is there any situation to avoid the `NoSchedule` machine pools for edge workloads?

Reviewer 6: @makentenza here

Reviewer: Is this asking if there's a way to avoid having the taint, or whether there's a use case for having the edge pools and worker pools run the same workloads unconditionally?

Author: > Is this asking if there's a way to avoid having the taint, or whether there's a use case for having the edge pools and worker pools run the same workloads unconditionally?

I mean if you'll think that using taint is the best approach for edge pools based on all scenarios and limitations described in this EP.

Reviewer: IMO the taints/tolerations approach is the best and canonical way to achieve this, should we close this question?

respectively, to allow application developers to deliver low-latency applications
at scale to the end users.

As the resources in that locations were designed for specific workloads, and
also have limited and more expensive cloud services, cluster operators should be
able to create worker nodes on the edge networks on day zero operations.

This enhancement proposes a new machine pool named `edge`, which will create
Machine Sets for each availability zone on the AWS Local Zones to provision compute
resources with `NoSchedule` taint.

The EP is divided into three phases:

- Phase 0: Research. Understand the Local Zones using the OKD/OCP, create a use case, and
  provides the documentation of the installation of a cluster in an existing VPC.
- Phase 1: Install a cluster in user provided network introducing to the installer a new
  machine pool named `edge`. The MachineSet manifests will be created for each subnet
  provided on `platform.aws.subnets` for Local Zones. The Machine will be protected with
  taints to avoid running the normal platform workloads.
- Phase 2: The installer should automate the network resource creation when the user

>   provides zone names with the type `local-zone` on the `platform.aws.zones` or

Reviewer 5: > `platform.aws.zones`

nit: The `zones` field is only in the machine pools. There is a `defaultMachinePlatform` machine pool in `platform.aws` so there is `platform.aws.defaultMachinePlatform.zones`, but that would imply the zone would be used for both control plane and compute.

My understanding is we want to limit these edge zones to compute, so I think this should be:
```suggestion
  provides zone names with the type `local-zone` on the 
```

Author: @patrickdillon thanks for getting this. That's correct, it's not a goal to use `platform.aws.defaultMachinePlatform.zones` for `edge` compute pools.

from NYC to node hosted in Local Zones in Miami, instead of New Year node).

As the applications using edge locations are very specific workloads, cluster admins
want to have available the compute nodes on the board without mixing cluster components
on those pools, and would like to specify which workloads will run for each location to
optimize the cloud resources. For that reason, the nodes on the edge pool have taints to
`NoSchedule`.

The installer supports day 2 operation in Local Zones subnets following these conditions:
- the VPC must have enough free CIDR blocks to create subnets in Local Zones
- the Local Zones group must be activated manually
- the subnets must be created manually
- the subnet must be associated with a route table
- the instance type offers must be checked, and the type is chosen for each Local Zone
- the MachineSets manifests must be created manually for each zone Local
- preferably, the node should not receive regular cluster workloads, mainly ingress. due the
  unsupported network load balancers, latency, and costs.

Presently, the installer does not support Local Zones subnets on day zero operations
for those reasons:

> - the Local Zones do not support network load balancers (NLB or CLB), used by

Author: Yes, Application Load Balancer at the moment. I asked AWS, and register a product interest, but they haven't plans (at least open for customers) to support network load balancers.

  so I can have a cost-effective infrastructure, avoiding mixing workloads in different locations.

- As a cluster admin, I would like to select existing subnets from the local and
  the parent region zones, to install a cluster, so I can manage my custom networks with
  my automation.

Phase 2:

- As a cluster admin, I would like to install OpenShift clusters, extending the
  compute nodes to the Local Zones in my day-zero operations without needing to
  set up the network and compute dependencies, so I can speed up the edge adoption
  in my organization using OKD/OCP.

### Goals

Phase 1:

- Admins can create machine pools named `edge` when installing a cluster in existing VPC with subnets in AWS Local Zones.
- Admins can select and configure subnets on Local Zones before cluster creation.
- Ensure the installer allows creating a new machine pool for `edge` workloads
- Ensure the installer can create the MachineSet with `NoSchedule` taints on edge machine pools.
- Ensure Local Zones subnets will not be used on `worker` machine pools or control planes.
- Ensure the Wavelength zone will not be used in any machine pool

> Phase 2:

Reviewer 5: In Phase 2, the installer is creating network resources... that does not include the application load balancers, right? The ALBs will always be created Day 2?

Author: `Network` I would mean every manual resource created by the user on Phase-1 to install a cluster in the existing network/VPC: VPC, subnets (regular and LZ), and dependencies.

Yes, ALB should be installed on Day 2, the user can opt to use it or not, depending on the use case they've chosen to use Local Zones.

I will clarify the second bullet to keep it more explicitly of resources Phase 2 will create. Thanks for pointing it out.

Author: @patrickdillon I described which network resources on this section. Thanks

will be calculated automatically on the install time. If admins would like to expand
they need to check if there's free CIDR to create subnets in the new locations, create
the subnets and dependencies, then create the MachineSet manually.
If there is no available CIDRs, they need to install a new cluster in existing
subnets, and manage their own subnet CIDR blocks, creating the new locations (subnets
and MachineSets) on Day-2.

The total block allocated to the subnet on Local Zones could be underutilized,
avoiding future expansions. This is a risk to the current IPI. A note on the documentation
can be added: "if there is a plan to expand the subnets, the installation in existing VPC
may be better".

New Local Zones locations can be created without any type on the preferred list, then the MachineSet
manifest generated will be created with an unsupported instance type, generating in `Failed`
instance. We can keep the list of regions and locations tested on the Documentation, so we can
provide better feedback on what is currently supported.

### Drawbacks

- Customers would prefer to deploy only on Day 2, or in existing VPC creating their own MachineSets

> - Compute nodes in the edge may end up idle if not correctly monitored

Author: I don't have enough understanding of how the AS can work on the edge pools, considering the taints. Can we have MachineAutoscalers objects for the `edge` MachineSets and then scale normally when the workloads with proper tolerations have been set?

> - Compute nodes in the edge may end up idle if not correctly monitored

Reviewer: That should work yep! The autoscaler is aware of taints and tolerations, node selectors etc so it should be able to work as intended

---
title: aws-custom-edge-machineset-local-zones
authors:
  - "@mtulio"
reviewers:
  - "@rvanderp3"
  - "@faermanj"
  - "@patrickdillon"
  - "@elmiko"
  - "@JoelSpeed"
  - "@sdodson"

> approvers:

Author: @patrickdillon @JoelSpeed  anyone else to add as an approver here?

  - "@patrickdillon"
  - "@JoelSpeed"
  - "@sdodson"
api-approvers:
  - None
creation-date: 2022-09-05
last-updated: 2022-09-05
tracking-link:
  - https://issues.redhat.com/browse/RFE-2782
see-also: []
replaces: []
superseded-by: []
---

# Allow Customer-Provisioned edge machine pool for AWS Local Zones

## Summary

Edge adoption is increasing and providers are launching many offerings that fill
that area. AWS launched the [Local Zones](https://aws.amazon.com/about-aws/global-infrastructure/localzones/)
and the [Wavelength](https://aws.amazon.com/wavelength/features/) zones to extend the
VPC network to the metropolitan area from big centers, and for Radio Access Network,
respectively, to allow application developers to deliver low-latency applications
at scale to the end users.

As the resources in that locations were designed for specific workloads, and
also have limited and more expensive cloud services, cluster operators should be
able to create worker nodes on the edge networks on day zero operations.

> This enhancement proposes a new machine pool named `edge`, which will create

Reviewer 5: Instead of asking users to set `installconfig.compute.name` to `edge`, would it make sense to validate/generate machinesets based on the `installconfig.compute.platform.aws.zones` field? Users would need to explicitly set the local zone in `installconfig.compute.platform.aws.zones` to use this functionality. It seems like `name` is more associated with the role of the node in the cluster, and would these still be considered "worker" nodes?

Author: Yes, it is possible. The name `edge` is used to improve the UX for the user when explicitly choosing the `edge` zones (Phase-2), and/or customizing the nodes in this pool, exclusively, without affecting the regular worker pool.

But when keeping the `edge` machine pool, I think there are other benefits of decreasing the implicit things on the installer, and the logic, when trying to fit the best compute type for this pool. In [this thread](https://github.com/openshift/enhancements/pull/1232#discussion_r979980109), @JoelSpeed and I discussed the benefits of not keeping the same compute type in `edge` pool, other than the default in regular zones.

Having said that, I can see some cons when not using `edge` pools. For example:
- if the user inputs the compute type for `installconfig.compute[name==worker].platform.aws.type` to an invalid instance type of Local Zones (which is very limited), the installer should be failed
- if the user input the compute type for `installconfig.compute[name==worker].platform.aws.type` to a valid instance type of Local Zones, the installer will deploy a more expensive cluster and will, considering current pool available in most zones, return the types used by the installer in 1+ years to use 5th-gen of instances, as the 6th-gen, default, is not generally available.
- if the user input to use `gp3` on the config, the LZ worker nodes will silently (to the installer) fail when MAPI will provision it when operating in zones that do not have this support (there is not API to query if an given zone has support for X type of EBS). If set gp2, we will use slower and more expensive block storage for regular worker nodes.
- if the user would like to deploy a subnet in the LZ with a single compute pool (worker) in Phase-2, they will need to provide the LZ zone name in `installconfig.compute[name==worker].platform.aws.zones[]`, but what will be the new default behavior of installer in this case if only LZs are defined? Currently, it will fail, otherwise, they will need to provide each regular AZs + LZs name. I think it will also impact the UX as they just want to extend to the LZ the regular IPI.

In my point of view, keeping the edge pool available to the users could provide more transparency to them in what is being deployed.

Author: > One thing we should keep in mind is that the idea of an edge machinepool is platform agnostic, so the idea should generally make sense across platforms. I do believe this has already been covered in this enhancement but just wanted to state it as a reminder.

Yes, it is.

Thanks for pointing it out. I am closing this discussion.

Machine Sets for each availability zone on the AWS Local Zones to provision compute

> resources with `NoSchedule` taint.

Reviewer 2: Are we sure that we'll always want this?

Author: @sdodson let me reference and summarized previously discussions about this topic:

- [here](https://github.com/openshift/enhancements/pull/1232#discussion_r1045918478) I wrote, and I closed, the `Open Question` about this.
- [here](https://github.com/openshift/enhancements/pull/1232#discussion_r979980109) I detailed the reasons to create `NoSchedule`taints for "edge" nodes. TL;DR: users should have more control in which workload will run on the edge as it is more expensive and can impact the latency when serving requests from/to different locations that the edge is deployed (this is the main feature when choosing to deploy those nodes: achieve low latency to their clients), and shouldn't be balanced as a normal nodes in the parent region (regular AZs). [Here](https://github.com/openshift/enhancements/pull/1232#discussion_r1000204815) I am summarizing the simple benchmark of routing requests from/to 'undesired' locations.

Let me know if you have any questions or thoughts in reasons to not use it.

Reviewer 2: Yeah, I saw the rationale that it's more expensive and I assume that's likely to always be the case given cloud economics.

I'm just not sure we wouldn't potentially have a situation where a customer wants to operate with all workload on localzones with no normal zone workers present. Is that a compact cluster with schedulable control plane nodes with an edge pool or would there be scenarios where they prefer to run infrastructure components like Image Registry localized to the edge pool?

Author: Certainly, we can have many variants to use the AWS edge computing (and when [I joined sessions](https://cvws.icloud-content.com/S/AcFiavBeyXTI9iglWF30s9_YTU5m/IMG_0106.JPG?o=AtZsXb_vS24yNb6g1BbMbYs87wYtEb4GWl4QFKoHfQJd&v=1&z=https%3A%2F%2Fp42-content.icloud.com%3A443&x=1&a=CAogt9Lz05vQtdQYZ5M4ksMcnPPGdt_neLC42WPkUMOkgqcSaRDTkuuS2DAY06n-l9gwIgEAUgTYTU5maiclXEKoBCQjU1tsIijqHd15FiKXIa3IOxOZcJM1eLITmYbcGnsJ6xJyJ7kZCXVVM-B2XzCGa1RWX6wXI8cKwL0LfPxmdepI8zjI66DFYjWpxw&e=1672940065&r=55be105d-b36f-4b1f-b985-9c49052863d2-46&s=DN1DAAAijKjP8ugbI8wCkI1fhNw) at AWS re:Invent for this topic of what customers are already doing in the edge, it's really very broad), but most of them could require an extra configuration/customization which I think would be good to choose other install methods than the covered in this EP for the installer using IPI, for example:

- Disconnect clusters may require an extra configuration (associate the LZ subnet into one of the parent's zone route tables to use Nat Gateway there), and will not be covered by this EP
- Some customers operating in countries that do not have regular zones, only edge zones (like Buenos Aires/Argentine, recently announced) may want to data sovereign and low latency deploying the entire cluster in those zones. But Control Planes will not be supported in the Local Zones due to the NLB limitation.
- If no zone workers are present in the regular zones, and edge hasn't NoSchedule taints, the default ingress deployment will fail due to the limitation of NLB/CLB, failing the Console too, and other apps required by it.
- I think they can choose to deploy image-registry in the edge, for example serving the apps there and, maybe their on-prem infra when they have Direct Connect with their infra close to the edge location, but I think it's very specific and can be achieved by [post-installation tasks, as usually do with infra nodes](https://docs.openshift.com/container-platform/4.11/post_installation_configuration/cluster-tasks.html#post-install-creating-infrastructure-machinesets-production)
- Deploying cluster image-registry to the edge, operating in more than one zone groups could imply in latency much higher when serving images inter-edge zones.

So, it seems to be hard to keep a supported cluster in Local Zones without, at least, infra and control planes in the parent region.

Open to hear what do you think, if we may need to cover more scenarios, etc

The resources on Local Zones were designed for applications delivering low latency to
the end-users, in general, hosted in the data plane nodes. Furthermore, the price of
cloud services is more expensive, and the services are limited. Finally, mixing the
ingress router backends from different locations in the Local Zones can impact
drastically, and directly, the end users' experience (ex: routing requests from users
from NYC to node hosted in Local Zones in Miami, instead of New York node).

As the applications using edge locations are very specific workloads, cluster admins
want to have available the compute nodes on the board without mixing cluster components
on those pools, and would like to specify which workloads will run for each location to
optimize the cloud resources. For that reason, the nodes on the edge pool have taints to
`NoSchedule`.

The installer supports day 2 operation in Local Zones subnets following these conditions:
- the VPC must have enough free CIDR blocks to create subnets in Local Zones
- the Local Zones group must be activated manually
- the subnets must be created manually

> - the subnet must be associated with a route table

Reviewer 2: This routing allows both a local zone worker to connect to `api-int` load balancer and for control plane hosts to connect to the kubelet on local zone workers?

Author: Yes, it allows. It is all in the same broadcast domain, with same Security Group as regular workers.

- the instance type offers must be checked, and the type is chosen for each Local Zone
- the MachineSets manifests must be created manually for each zone Local
- preferably, the node should not receive regular cluster workloads, mainly ingress. due the
  unsupported network load balancers, latency, and costs.

Presently, the installer does not support Local Zones subnets on day zero operations
for those reasons:
- the Local Zones do not support network load balancers (NLB or CLB), used by
  ingress deployed on the `worker` pool.
- the Local Zones do not support Nat Gateway, used as a gateway on the default IPI

>   deployment for each subnet and availability zone.

Reviewer 2: Do we know if they plan NAT gateway support?

Author: I asked, no plans.

- the default instance type for compute nodes used on IPI, `m6i.xlarge`, is not
  available on Local Zones

> - the default EBS type for IPI, `gp3`, is not available in the most Local Zones locations

Reviewer 7: cluster-storage-operator uses `gp3` as the default storage class. Since most Local Zones can't use such a volume, is there an expecation that OCP is smart enough to use `gp2` as the default for PVs? If so, cluster-storage-operator needs to be changed to deploy gp2 as the default (that's easy) when local zones are detected (that could be hard, is such info available in any API object?)

Author: @jsafrane, thanks for pointing it out. Could you please review if this sentence is correct?
- The default StorageClass will be used only when the PV is required. The PV will be used on Local Zones nodes only when users choose to deploy workloads there - due to the taints on those nodes. The users should create tolerations on their deployment spec to run workloads on the edge on day-2, in that steps they should define the `storageClass`=`gp2-csi` when the specific Local Zone location does not support gp3.

Reviewer 7: ack, that sounds good. User is responsible for using the right storage class when a workload is targeting node in a local zone.

Author: mentioning this discussion to the [specific EBS section](https://github.com/openshift/enhancements/pull/1232/commits/8b96adf7661e73f0b8b76e9c2e9c5185093df361)

  my automation.

- As a developer of VR/AR apps, I would like to use the power of Cloud GPU instances to process
  images closest to my users, so I can deliver real-time experience.

- As an autonomous application running in an embedded machine, I would like to process the data collected from many
  sensors and cameras on the Edge of the Cloud with real-time decision-making to my production line.

Phase 2:

- As a cluster admin, I would like to install OpenShift clusters, extending the
  compute nodes to the Local Zones in my day-zero operations without needing to
  set up the network and compute dependencies, so I can speed up the edge adoption
  in my organization using OKD/OCP.

### Goals

Phase 1:

- Admins can create machine pools named `edge` when installing a cluster in existing VPC with subnets in AWS Local Zones.
- Admins can select and configure subnets on Local Zones before cluster creation.
- Ensure the installer allows creating a new machine pool for `edge` workloads
- Ensure the installer can create the MachineSet with `NoSchedule` taints on edge machine pools.
- Ensure Local Zones subnets will not be used on `worker` machine pools or control planes.

> - Ensure the Wavelength zone will not be used in any machine pool

Reviewer 5: How do we distinguish wavelength zones?

Author: The [API attribute](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AvailabilityZone.html)  `zoneType==wavelength-zone`

volume type for Local Zones, `edge` machine pool, will be fixed to `gp2`.

The EBS Volume Type can be changed on the `edge` machine pool option
`compute[name=edge].platform.aws.rootVolume.type` on the `install-config.yaml`.

#### edge MachineSets

The installer must generate one MachineSet for each Local Zones location, considering
all the limitations on the Local Zones (instance types, EBS, network, etc).

The MachineSet for the `edge` pool will also set those custom labels:

- `node-role.kubernetes.io/edge=''`: used by `taint` and to quickly identify on `oc get nodes`
- `location=local-zone`: can be used by customers to deploy custom applications across all Local Zones nodes
- `zone_group=<zone_group_name>`: can be used by the developer to deploy applications in custom locations

The MachineSet also must have the taint to `NoSchedule` to the label `node-role.kubernetes.io/edge=''`.
This taint will prevent to schedule of regular cluster workloads (router, logging, monitoring)
into distant geo-locations (Local Zones), allowing more control to the user to
choose the workloads running for `edge` location, and also more control of the costs.

#### General considerations

> The following items must be satisfied in this phase:

Reviewer 5: I think we should also state that any zones specified in the edge pool `compute[?name=='edge'].platform.aws.zones` must have a corresponding subnet declared in `platform.aws.subnets`. This should also be true for phase 2.

That is, the following config would be invalid in both phase 1 & phase 2, because there is no supplied subnet for the zone specified in the `edge` pool (in phase 2, the installer will not create missing subnets when a user supplies existing subnets).

```yaml
apiVersion: v1
baseDomain: example.com
metadata:
  name: edge-cluster
compute:
- name: edge
  platform:
    aws:
      zones:
      - us-east-1-mia-1a
platform:
  aws:
    region: us-east-1
    subnets:
    - subnetID-us-east-1a
    - subnetID-us-east-1b
    - subnetID-us-east-1c
    - subnetID-us-east-1-nyc-1a
pullSecret: '{"auths": ...}'
sshKey: ssh-ed25519 AAAA...
```

Author: @patrickdillon thanks for pointing it out. I will make sure this EP and implementation covers it.

The internal data transfer costs (within the VPC: kubelet<->NLBs) will be the same
of the other (normal workers) nodes[1][2].

> *Data Transfer charges in AWS Local Zones is the same as in the Availability Zones today. Learn more about data transfer prices "in" and "out" of Amazon EC2 here.*

The compute resource in Local Zones is more expensive, for example, `m5.xlarge` hosted
in the parent zones of `us-west-2` will cost `$0.192/hr`, and in the LAX (Los Angeles)
Local Zones is `$0.23/hr`, almost 17% more expensive than the regular zone.

When following the architecture of using private subnets in Local Zones re-using Nat Gateway
from the parent region (proposal #2 in the last section), considering NAT Gateway
currently is not supported on the Local Zones locations, the following fees will be
applied when the node needs access to the internet:

- Fee for data transferred "in" to and "out" from NAT GW located in another AZ (`$0.01/GB` for each direction)
- NAT GW fee per GB processed `$0.045`

There is another cost of `$0.045/hr` for NAT GW allocation, which is not exclusive to a node in
Local Zones as the Nat Gateway is already created to be used in a regular zone.

> When the node is in the public subnet (described in this EP), the fees applied will be the same to access the internet.

Reviewer: This section is really great, would be good to include a real world example if that's something we have to hand, I'm thinking something like
"Given a workload that is processing x ingress data and y egress data, this would cost x on a normal zone and y on an edge zone", I don't know if we have some average data or whatever that we can suggest for a real world example though

Author: @JoelSpeed thanks.
I added one more paragraph to the "Infrastructure Costs" section trying to address and provide enough resources to address your feedback. I tried to be succinct providing resources to deepen the calculation of the scenarios as I was taking time to build many different hypothetical scenarios without enough data from real customers' use cases to enrich the content.

Author: https://github.com/openshift/enhancements/pull/1232/commits/7a2c770ce472992129114c223016379a80bd3f13

#### User-workload deployments

Application developers must set the tolerations for the application running for each location.

The pod spec example below describes how to deploy one application on the node on the zone group `ZONE_GROUP`:

```yaml
      nodeSelector:
        zone_group: ZONE_GROUP
      tolerations:
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
```

<!--
> TODO create one example of deploying the application on the same ZONE_GROUP, spreading it into different zones.

> Currently only Los Angeles (LAX) location has more than one zone[1];

The example below shows how to deploy one application with two replicas distributed
between two Local Zones[1] in the same zone group[2].

[1] Local Zones available in Los Angeles `us-west-2-lax-1a` and `us-west-2-lax-1b`
[2] Local Zones group of Los Angeles `us-west-2-lax-1` -->

#### User-workload ingress traffic

To expose the applications to the internet on AWS Local Zones, application developers
must expose their applications using AWS Application Load Balancers (ALB). The

> [ALB Operator](https://docs.openshift.com/container-platform/4.11/networking/aws_load_balancer_operator/install-aws-load-balancer-operator.html) is available through OLM on 4.11+.

Reviewer: Does the service object not support ALB when implemented by the AWS cloud provider?

Author: Yeah, AFAIK the AWS cloud provider supports only network load balancers (CLB/NLB) natively. ALB is handled by [AWS Load Balancer Controler](https://kubernetes-sigs.github.io/aws-load-balancer-controller).

- Server: Location of the server is running the sample application
- Server `Main`: Any Compute machine running in the Parent region (us-east-1)
- Server `NYC`: OCP Compute Machine running in AWS Local Zone located in New York
- Server `MIA`: OCP Compute Machine running in AWS Local Zone located in Miami

Highlight the time to Connection Time (from a client in NY): The time to connect
to the NYC zone was 3x faster than the parent region, and 10x faster than the
location far from the user:

```text
From NY -> To Parent Region Zone nodes (us-east-1)
      time_connect:  0.009792 Sec
                   ----------
        time_total:  0.022362 Sec

From NY -> To NY node
      time_connect:  0.003788 Sec
                   ----------
        time_total:  0.015983 Sec

From NY -> To Miami node
      time_connect:  0.032383 Sec
                   ----------
        time_total:  0.089729 Sec
```

You can find the raw files [here](https://github.com/mtulio/mtulio.labs/tree/master/labs/go-geo-app/data-benchmark).

<!-- #### Communication Between Instances in different zone groups

> TODO: check if there is any issue, it was mentioned by LZ experts. -->

### Risks and Mitigations

> Administrators may want to add new locations on clusters already created by the instlaller.

Reviewer 8: nit: "instlaller" -> "installer"

1. Is there any situation to avoid the `NoSchedule` machine pools for edge workloads?

Phase 2:

1. The Local Zones currently have limitations of Nat Gateways, which keep the gateway
  strategy more complex to provide egress traffic for compute nodes deployed on
  that zone. Can we keep the edge nodes in the public subnet? Do we need to
  create a dedicated security group instead of using/sharing with the default worker
  machine pool?

1. How we can keep extending the Zones for eventually new locations? AWS has a fixed
  number of Local Zones available, when we create a cluster we will try to calculate
  the subnet IPs to cover all IP ranges allowed by VPC (/16) depending on the number
  of subnets available. I also wonder if we can consider a smaller range of IPs for
  subnets operating in Local Zones, allowing the customer to grow it later. Can we
  consider using a lower CIDR for subnets hosted in Local Zones locations?

1. How we can advise customers to grow locations on Day 2? Can we keep a product
  documentation page with that instructions? A sort of instructions on how to create
  CloudFormation Stack to create a new zone for the new location that is available, or

>   the customer wants to support in the future.

Reviewer 8: This is also likely to be useful for HyperShift, where... something? is creating subnets and such...  Although I'm not sure if that's on-demand at day-2, or something that happens at install-time and is assumed to persist throughout the cluster's lifetime.  It looks like [the NodePool's `subnet` property is an optional pointer][1], which hints at having some HyperShift-y logic for day-2 management.  But having Go installer code, instead of Terraform, that makes it easy to shift this into a day-2 controller at some point might be worthwhile.  Or maybe I'm overthinking this...

[1]: https://github.com/openshift/hypershift/blob/4073d66298251f2c7bb790c5be4afbec6af4dc28/api/v1beta1/nodepool_types.go#L569-L614

Author: I think expanding zones will be very common, mainly when we consider that Local Zones are quite new and are expanding around the globe continuously, and customers would like to use new locations in existing clusters but will be blocked as the installer consumed all the CIDRs blocks from the VPC when creating the subnets. I raised this question to the Arch call today, it's not clear if we can customize the CIDRs blocks of subnets on install time.

I don't have the data about *what users want*, but for me, it seems to be a very common day-2 operation, mainly when they want to create a smaller cluster and expand it continuously over time. I also think that we don't need to allocate big CIDR blocks for edge zones, for example, if the customer would like to be present in many edge zones within a given region just to deploy a few nodes there - why add a `/22`for a subnet which expects a few assets?! It's complex use cases, IMHO. Maybe it would be better to tell users to install in existing VPC and allow them to manage with their IPAM and expand in their end/automation, when the new location/zone should be added, they will create it on day 2 (network items and MachineSets).

Author: Section [`VPC CIDR block (Machine Network)`](https://github.com/openshift/enhancements/pull/1232/files#diff-f2590494c9138551ab5b3d5632d11170fe8742a128256f9e1d8ebc63baa19f41R953-R995) added to the `Risks and Mitigations` detailing how to terraform will behave when deploying a cluster to prevent CIDR exhaustion, allowing subnet expansion on day 2.

#### Dev Preview -> Tech Preview
#### Tech Preview -> GA

N/A

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

N/A. The resource creation, Local Zones subnets, and MachineSets in the edge
machine pool are available only in day zero operations. All the further
lifecycle will be done as a normal machine.

### Version Skew Strategy

N/A. The same as `Upgrade / Downgrade Strategy`.

### Operational Aspects of API Extensions

N/A.

#### Failure Modes

N/A.

#### Support Procedures

N/A.

## Implementation History

Currently, a few research documents were created to understand how the edge networks
of AWS can be used on OKD/OpenShift, most of them were focused on extending current clusters
using MachineSets when installing a cluster in existing networks.

The first research was focused on Day-2 operations, how to extend current IPI deployment
to Local Zones subnets, created separately, and how to use the workloads on it, since the
openshift router supports only network load balancers (NLB/CLB), which is not supported
on Local Zones.

## Alternatives

- Cluster admins can create subnets and the MachineSet manifest at the day zero operations

>   using the installer option `create manifest`, then trigger the cluster creation.

Reviewer 8: Probably worth explicitly explaining why these alternatives were not chosen.  Something like:

> But manually creating subnets and MachineSets is tedious and error-prone.  And there are no in-cluster operators to assist with these changes post-install.  So we're sticking the helper in the installer to deliver user convenience with minimal developer effort.

or whatever the motivation for avoiding the alternatives actually is.

Reviewer 8: [Looks like][1] the `subnet` pointer is just about allowing HostedCluster lookups, not about day-2 creation when adding a NodePool.  But that's today, and I still think it would be nice to have active controllers for this infra at some point.  But folks may disagree, and even if they agree with me, that active-controller point may be so far out in the distant future that it's not worth planning around today.

[1]: https://github.com/openshift/hypershift/blob/4073d66298251f2c7bb790c5be4afbec6af4dc28/cmd/nodepool/aws/create.go#L53-L59

Author: > Probably worth explicitly explaining why these alternatives were not chosen. Something like:
> 
> > But manually creating subnets and MachineSets is tedious and error-prone.  And there are no in-cluster operators to assist with these changes post-install.  So we're sticking the helper in the installer to deliver user convenience with minimal developer effort.
> 
> or whatever the motivation for avoiding the alternatives actually is.

Good point, thanks for raising it. I will update you with that suggestions. In the Motivation section, I put the long-list of tasks to do, pointing this with your suggestions *sgtm*.

  images closest to my users, so I can deliver real-time experience.

- As an autonomous application running in an embedded machine, I would like to process the data collected from many
  sensors and cameras on the Edge of the Cloud with real-time decision-making to my production line.

Phase 2:

- As a cluster admin, I would like to install OpenShift clusters, extending the
  compute nodes to the Local Zones in my day-zero operations without needing to
  set up the network and compute dependencies, so I can speed up the edge adoption
  in my organization using OKD/OCP.

### Goals

Phase 1:

- Admins can create machine pools named `edge` when installing a cluster in existing VPC with subnets in AWS Local Zones.
- Admins can select and configure subnets on Local Zones before cluster creation.
- Ensure the installer allows creating a new machine pool for `edge` workloads
- Ensure the installer can create the MachineSet with `NoSchedule` taints on edge machine pools.
- Ensure Local Zones subnets will not be used on `worker` machine pools or control planes.
- Ensure the Wavelength zone will not be used in any machine pool

Phase 2:

> - Admins can configure the Local availability zones before cluster creation.

Author: ```suggestion
- Admins can configure the Local Zone names on install-config.yaml, `compute[?name=="edge"].platform.aws.zones`.
```

With that strategy, all the current GA locations will be covered with supported instance
types, and when the most recent instances (currently 6th-gen `m6i.xlarge`) will be
offered on those locations, it will be selected automatically without needing to
change the installer.

The logic of the preferred instance type will be replaced for the `edge` pool when
the user provides the instance type on the install-config, attribute `compute[name=edge].platform.aws.type`.

#### Volume type

The EBS Volume Type `gp3`, used by default on IPI, is not broadly available
across all GA Local Zones locations. [The Local Zones Features page](https://aws.amazon.com/about-aws/global-infrastructure/localzones/features) is the current place to check which location supports `gp3`.

AWS also does not provide an API to query the zone offering for EBS flavors. The EBS
volume type for Local Zones, `edge` machine pool, will be fixed to `gp2`.

The EBS Volume Type can be changed on the `edge` machine pool option
`compute[name=edge].platform.aws.rootVolume.type` on the `install-config.yaml`.

If the user-workloads requires an Phisical Volume, the `storageClass` can be set to `gp2-csi`

> on the deployment config.

Reviewer 7: What's "Phisical Volume"? Do you mean PersistentVolume Kubernetes object?

Author: Yes, typo. Thanks for getting that. I will fix it asap.
We are working to add this[1] to the documentation to help users and inform them about the limitations of [2]. Open to suggestions on that user-facing doc too. Thanks

[1] https://github.com/openshift/openshift-docs/pull/57063
[2] https://github.com/openshift/openshift-docs/pull/57736

Author: Documentation is live on 4.12+ on the Cluster tasks exemplifying workloads deployment in Local Zone nodes, including PV volumes:
https://docs.openshift.com/container-platform/4.12/post_installation_configuration/cluster-tasks.html#installation-extend-edge-nodes-aws-local-zones_post-install-cluster-tasks

```yaml
      nodeSelector:
        machine.openshift.com/zone-group: <zoneGroupName>
      tolerations:
      - key: "node-role.kubernetes.io/edge"
        operator: "Equal"
        value: ""
        effect: "NoSchedule"
```

#### User-workload ingress traffic

To expose the applications to the internet on AWS Local Zones, application developers
must expose their applications using AWS Application Load Balancers (ALB). The
[ALB Operator](https://docs.openshift.com/container-platform/4.11/networking/aws_load_balancer_operator/install-aws-load-balancer-operator.html) is available through OLM on 4.11+.

To explore the best of deploying applications on the edge locations, at least one new
ALB `Ingress` must be provisioned by location to expose the services deployed on the
zones.

If the cluster admin decides to share the ALB `Ingress` subnets between different locations,
it will impact drastically the latency for the end-users when the traffic is routed to
backends (compute nodes) placed in different zones that the traffic entered by Ingress.

The ALB deployment is not covered by this enhancement.

#### Benchmark results

> 

Reviewer 9: This section is great.
